cloud/store/migration.go

```diff
@@ -0,0 +1,179 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package store
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	schemaTypes "github.com/weaviate/weaviate/adapters/repos/schema/types"
+	cmd "github.com/weaviate/weaviate/cloud/proto/cluster"
+)
+
+const (
+	raftMigrationFileName        = "raft_migration_progress.json"
+	raftMigrationFilePermissions = 0o644
+)
+
+type migrationProgress struct {
+	MigratedClasses map[string]interface{} `json:"migrated_classes"`
+}
+
+func NewMigrationProgress() *migrationProgress {
+	return &migrationProgress{MigratedClasses: make(map[string]interface{})}
+}
+
+// MigrateToRaft will perform the migration from schemaRepo to the RAFT based representation stored in st.
+// If the current node is not the leader, the function will return early without error.
+// If the current node is the leader, it will read the schema from schemaRepo and import it in the RAFT based
+// representation using st.Execute to ensure RAFT has log entries for the schema.
+// It returns an error if any happens during the migration process.
+func (st *Store) MigrateToRaft(schemaRepo schemaTypes.SchemaRepo) error {
```

Comment:
I disagree on a few points.

> Given that user action might be unpredictable (e.g. falling back to a non-raft version), there is a risk of missing of data lost if migration halts midway

We should make it clear the downgrading between 1.25 and <1.25 leads to the risk of schema loss and encourage our users to have backups at the ready. Note we will not lose data as we have the same underlying storage for the DB, just the schema.

> In case of failure there is no guarantee that the same server will execute the migration. This could lead to clashes of update between nodes.

The helm chart should ensure the same pod & therefore the same volume will restarted/retried so we will have the progress stored on disk. Also we discussed in calls earlier that on the initial release for 1.25 we will have only one leader node in the helm chart. So only this node will take care of doing the migration.

> It fails to ensure a safe schema migration in case of failure

It has mechanisms to recover failure and try to reach a whole schema at the end of it. Which failures do you have in mind where we wouldn't be able to recover ?